packages/scripts/src/findFaq.ts

```diff
@@ -0,0 +1,139 @@
+  const conversationsCollection = db.collection<Conversation>("conversations");
+  const originalMessages = conversationsCollection.aggregate<
+    SomeMessage & { indexInConvo: number; convoId: ObjectId }
+  >([
+    {
+      $match: {
+        // Include only conversations that actually had user input
+        "messages.role": "user",
+      },
+    },
+    {
+      // With replaceRoot below, pass each message from conversation to next
+      // stage in pipeline
+      $unwind: {
+        path: "$messages",
+        includeArrayIndex: "indexInConvo",
+      },
+    },
+    {
+      $addFields: {
+        "messages.indexInConvo": "$indexInConvo",
+        "messages.convoId": "$_id",
+      },
+    },
+    {
+      $replaceRoot: { newRoot: "$messages" },
+    },
+    { $sort: { convoId: 1, createdAt: 1 } },
+  ]);
+
+  // Combine user question and responses into individual question & response
+  // objects. In the original conversation object, every 'user' message is
+  // followed by one or more non-'user' question, which is the response.
+  let currentQuestion: Partial<QuestionAndResponses> | undefined;
+  const questions: QuestionAndResponses[] = [];
+  const addQuestionToList = (
+    partialQuestion: Partial<QuestionAndResponses> | undefined
+  ) => {
+    if (partialQuestion === undefined) {
+      return;
+    }
+    const { embedding, question, responses } = partialQuestion;
+    assert(
+      embedding !== undefined &&
+        question !== undefined &&
+        responses !== undefined
+    );
+    questions.push({ embedding, question, responses });
+  };
+  for await (const message of originalMessages) {
+    switch (message.role) {
+      case "user":
+        {
+          addQuestionToList(currentQuestion);
+          currentQuestion = undefined;
+
+          const { embedding, content: question } = message;
+          if (embedding === undefined) {
+            // Earlier versions did not store user question embedding
+            continue;
+          }
+
+          currentQuestion = {
+            embedding,
+            question,
+            responses: [],
+          };
+        }
+        break;
+      case "assistant":
+      case "function":
+        {
+          currentQuestion?.responses?.push(message);
+        }
+        break;
+      default:
+        continue;
+    }
+  }
+  addQuestionToList(currentQuestion);
+  currentQuestion = undefined;
+
+  const { clusters, noise } = clusterize(
+    questions,
+    (q) => {
+      assert(q.embedding);
+      return q.embedding;
+    },
+    clusterizeOptions
+  );
+
+  const faqEntries = clusters
+    .map((cluster) => {
+      const embeddings = cluster.map(({ embedding }) => embedding);
+      return {
+        embedding: findCentroid(embeddings),
+        embeddings,
+        questions: cluster.map(({ question }) => question),
+        responses: cluster.map(({ responses }) => responses),
+        faqScore: cluster.length / (questions.length - noise.length),
```

Comment:
what is this metric?

Reply:
This should the relative frequency of the question cluster

Reply:
https://docs.google.com/document/d/1YQ8fcMneOPZGHbrowrKVEi-oHB82LnEtOlSuRJzsxRg/edit#heading=h.mbb4punhvs05

Reply:
https://docs.google.com/document/d/1_uaiiVyNSQvTrVm9VxSZMfz590voNAf73PgYyckwOX0/edit#heading=h.vsfxxnuz34oo

Reply:
ah ok. why `questions.length - noise.length` instead of `questions.length`? why remove the cluster-less "noise" questions from the score?

Reply:
Hmmm I had original just `/ questions.length` but then I liked the numbers I was getting without the noise more. The noise questions don't get saved in the faq collection, so it seemed sensible not to represent them in the faq score. That way you can take every entry from a findFaq run and sum all the faqScores to get 1. But it's completely arbitrary, I'm open to putting it back if you feel otherwise.

Reply:
the comparative frequency of clusters won't differ regardless of if we include `- noise.length` since noise.length is constant for all clusters in a findFaq() run, so i don't think it particularly matter which route we go. but with that being said, i personally prefer `cluster.length / questions.length` since it's simpler, and the idea of numQsInCluster/TotNumQs logically seems more straightforward

Reply:
Yeah, I think it might be nice to have faqScore include the invisible NON-'faq' questions... so I'll put it back the way it was!